# 2020-05-18 Experimental Redesign

See [../hpcc/2020-04-23/README.md](../hpcc/2020-04-23/README.md) for justification for redesigning our
experiments.

Overall goals:

- [high-priority] Test hypothesis: a changing environment promotes gene spread
  - Measure gene overlap/number of coding sites at a range of environment change rates
- [high-priority] Test hypothesis: high mutation rates promote gene overlap
  - Measure gene overlap/number of coding sites at a range of mutation rates
- [high-priority] Test hypothesis: more modular genomes are more evolvable than less modular genomes
  - Positive correlation between number of coding sites and final fitness?
  - More modular hand-designed architectures achieve higher fitness than less modular?
    - Can we also show that more overlapping architectures survive better in high mutation rate environments?
  - Environment transplant experiments - take modular architectures from changing environment, put in
    static environment. Do modular architectures (from changing environment) achieve higher fitness
    than organisms originally evolved in the static environment?
- Characterize: mutation-type (substitution vs gene movement vs insertions vs deletions) along successful
  lineages
  - Are different types of mutations more or less prevalent than expected (use random drift controls
    to establish a baseline)? Can we identify why?
- Verify robustness of results
  - Pivot off main parameters

Types of experiments

## Questions

- Mutation
  - Should we move forward with insertion/deletion & gene movement rates identified with original exploratory
    experiments? Or, should we re-sweep?
- Random drift representative organism
  - Currently still max fit (among the random genomes), should actually be random?

## Plans

Relevant parameter ranges

- Mutation rates
  - Bit flip (substitution) mutation rate(s): 0.0001, 0.0003, 0.001, 0.003, 0.01, 0.03, 0.1
  - Gene movement mutation rate(s)
    - Pivot rate: 0.003 (identified as reasonable with exploratory runs)
    - TODO (supplement) - characterize impact of increasing/decreasing (one step up/down)
  - Insertion/deletion rate(s)
    - Pivot rate: 0.001 (identified as reasonable with exploratory runs)
    - TODO (supplement) - characterize impact of increasing/decreasing (one step up/down)
- Environment change rates (differs between NK and gradient fitness models)
  - ~ changes/generation: **0**, 1/256, 1/128, 1/64, 1/32, 1/16, 1/8, **1/4**, 1/2, 1, 2/1, 4/1, 8/1, 16/1
  -  Pivot rates (?): 0, 1/4
  - NOTE - Cut a few of the higher change rates because they're less informative (environment is nonsense)
- Fitness models: NK, gradient
- Genetic architecture
  - Number of genes (flex max size appropriately): 8, 16, 32
  - Starting genome size: 32, 64, 128
  - Gene size: 8
- Selection
  - Drift vs. no drift
  - Use drift as a control for establishing baseline expectations.
    - baselines: coding sites, neutral sites, length, overlap, fitness(?)

Experiment by hypothesis

- (1) Test hypothesis: a changing environment promotes gene spread
  - Measure gene overlap/number of coding sites at a range of environment change rates
  - Experiment (15 x 2 = 30 conditions):
    - Change rate (approx) (x15): 0, 1/256, 1/128, 1/64, 1/32, 1/16, 1/8, 1/4, 1/2, 1, 2/1, 4/1, 8/1, 16/1, drift
    - Fitness model (x2): NK, Gradient
    - Selection: tournament size 8, 1 (for drift control => drift requires no change because environment is meaningless)
    - Bit flip rate: 0.003
    - Comparisons
      - (1) Number of coding sites across change rates (+ drift)
        - Expectation: higher change rates (to a point) have less overlap
      - (2) Fitnesses after environment lock down across change rates (+ drift)
        - Expectation: more modular genomes have higher fitness
      - (3) Fitness (after lock-down) vs. modularity comparison
    - To verify result robustness (later): pick a few change rates, vary off pivots
- (2) Test hypothesis: high mutation rates (relative to lower mutation rates) promote gene
  overlap
  - Measure gene overlap/number of coding sites at a range of mutation rates
  - Experiment (7 x 5 x 2 = 70 conditions):
    - Bit flip rate (x7): 0.0003, 0.001, 0.003, 0.01, 0.03, 0.1
    - Change rate (approx) (x5): 0, 1/128, 1/4, 4, drift
    - Fitness model (x2): NK, gradient
    - Selection: tournament size 8, 1 (for drift)
  - To verify result robustness (later): vary off a few pivots
- (3) Environment transplant experiments
  - Experiment
    - Evolve: static environment (x200), 1/4 changing environment (x100), drift environment (x100)
      - Take architectures from 1/4 changing and drift, evolve in static environments
      - Compare: static @ 60k, 1/4 changing in 10k static, drift in 10k static
- (4) Hand-designed architectures
  - Enumerate all possible architectures (for small gene count, gene size, and genome size)
    - Characterize distribution of graph properties
    - Evolve each for 10k generations in random static environment (x10)
    - Look at landscape (graph properties vs. fitness)
- (4.5) Select cross section for phylogeny tracking/analysis
  - Mutation distributions over time for successful lineages
  - Question: do different mutation rates/change rates have different distribution?
- (5) Verify robustness of results
  - Lower/higher gene counts - 8 and 32
  - Increase/decreased starting genome size - 32, 64, 256
  - Strength of selection - half/double
  - Increase/decreased gene movement rate - +/- half order of magnitude
  - Increased/decreased insertion/deletion rate - +/- half order of magnitude
- (6) Does environment similarity/fitness landscape similarity (w/respect to genes) predict
      overlap relationships?
  - Cursory look at data => not totally sure.
